---
output:
  bookdown::html_document2:
    toc: true
    toc_float: true
    toc_depth: 3
    number_sections: true
    theme: united
    highlight: tango
    df_print: paged
    code_folding: hide
    self_contained: false
    keep_md: false
    encoding: 'UTF-8'
    css: ./assets/course.css
---

```{r,echo=FALSE,message=FALSE,warning=FALSE,results='hide'}
## REPORT OPTIONS
## code relating to the report creation
## default working directory is the location of this document
## all code is run in the working directory as the root

# remove all variables
rm(list=ls())

# load libraries for document creation
library(knitr) # runs pandoc
library(bookdown) # for markdown
library(captioner) # for table and figure captions
#library(extrafont) # for custom font
#font_import(pattern="Gidole",prompt=FALSE) # font related windows
#loadfonts(device="win") # font related windows
# font used in plots
# basefont <- "Roboto"  # font related windows
# windowsFonts(sans=basefont)  # font related windows
# windowsFonts(serif=basefont)  # font related windows

# prepare captions
tc <- captioner::captioner(prefix="<b>Tab. </b>")
fc <- captioner::captioner(prefix="<b>Fig. </b>")

# set knit options
opts_knit$set(progress=TRUE,verbose=TRUE)
opts_chunk$set(dev="png",dev.args=list(type="cairo"),dpi=96,results="hold",fig.show="hold",fig.align="left",echo=TRUE,warning=FALSE,message=FALSE)
#options(knitr.table.format = "html") 
```

```{r,echo=FALSE,results='hide'}
## LIBRARIES AND VARIABLES
# only load the packages you need

# data handling
library(dplyr)
library(tidyr)
library(stringr)

# tables
library(kableExtra) # complete table
library(formattable) # table with conditional formatting
library(DT) # interactive table

# graphics
library(ggplot2) # static graphics

# interactive graphics
library(highcharter)
library(plotly)
library(ggiraph) # convert ggplot to interactive
library(dygraphs) # time series
library(networkD3) # network graph
library(leaflet) # interactive maps
library(crosstalk) # linking plots

# custom ggplot theme
theme_report <- function (basesize=12,font="Roboto") { 
    theme_bw(base_size=basesize, base_family=font) %+replace% 
        theme(
            panel.border=element_blank(),
            panel.grid.minor=element_blank(),
            panel.grid.major.x=element_blank(),
            legend.position="top",
            legend.direction="horizontal"
        )
}

#colours
rv_col_dark <- "#125687"
rv_col_light <- "#e7eef3"
rv_year <- 2018
```

<img src="./assets/logo_60.png" alt="logo_raukr" style="height:70px; position:absolute; top:1em; right:0; padding-right:40px; margin-top:7px">

<h3 class="toc-ignore rtitle">`r paste0("RaukR ",rv_year)`</h3>
<h1 class="toc-ignore rstitle">Introduction to Some Fancy Analyses</h1>
<h4 class="toc-ignore">`r format(Sys.Date(),format="%d-%b-%Y")`</h4>
<h4 class="toc-ignore">Your Name</h4>

---

<p class="abstract">
This is the course work material for RaukR `r rv_year`. Your can write general instructions here.
</p>

---

# R Markdown

* This is an R Markdown presentation. 
* Markdown is a simple formatting syntax for authoring HTML, PDF, and MS Word documents.
* This presentation is a demonstration of the capabilities of the R Markdown environment.
* In RStudio, you create a `.Rmd` text file, then click the **Knit** button.
<img src="./assets/knit-button.png" class="fancyimage">
* `.Rmd` documents can also be rendered as such:

```{r,echo=TRUE,eval=FALSE}
rmarkdown::render("report.Rmd",output_format="html_document")
rmarkdown::render("report.rmd",output_format="pdf_document")
```

# Text Formatting  

## Level 2 heading  
### Level 3 heading  
#### Level 4 heading  
##### Level 5 heading  
###### Level 6 heading 

*This is italic text*  
**This is bold text**  

An example of subscript: H~2~O  
An example of super script: 2^10^  
An example of footnote reference [^note1]  

[^note1]: That reference refers to this footnote.  

> This is a block quote. This
> paragraph has two lines.
>
> 1. This is a list inside a block quote.
> 2. Second item.

`This is inline code.`  

```
This is generic block code.
```

```{r,echo=TRUE,eval=FALSE}
This is R code.
```

<p class="vlarge">This is very large text.</p>
<p class="large">This is large text.</p>
This is normal text.
<p class="small">This is small text.</p>
<p class="vsmall">This is very small text.</p>
<p style="font-variant:small-caps;">This is small caps.</p>

# Layout  

Use `<div>content</div>` for block content and `<span>content</span>` for inline content. 

<p class="text-right" style="background-color: aliceblue">This content is aligned right.</p>
<p class="text-center" style="background-color: aliceblue">This content is aligned center.</p>

<p class="float-left" style="background-color: aliceblue">This content is pulled left.</p>
<p class="float-right text-center" style="background-color: aliceblue">This content is pulled right and text centered.</p>

<p class="float-left-small" style="background-color: aliceblue">This content is pulled left.</p>
<p class="float-right-large" style="background-color: aliceblue">This content is pulled right.</p>

<p class="float-left-large" style="background-color: aliceblue">This content is pulled left.</p>
<p class="float-right-small" style="background-color: aliceblue">This content is pulled right.</p>

## Links and Code  

Genome assembly and corresponding gene models `sacCer3/R64-1-1` were dowloaded from [http://www.yeastgenome.org](http://www.yeastgenome.org). TSS annotation `sgdGenes.sga` was downloaded from [http://ccg.vital-it.ch/mga/sacCer3/sgd/sgd.html](http://ccg.vital-it.ch/mga/sacCer3/sgd/sgd.html).

Following inspection of the data quality using `FastQC/0.11.2`, reads were trimmed off low quality bases and adapters using `Trimmomatic/0.32`.

```sh
java -jar trimmomatic.jar PE -threads 1 -phred33 \
$infile1 $infile2 $outfile1 $outfile1_npe $outfile2 $outfile2_npe \
ILLUMINACLIP:TruSeq3-PE.fa:2:30:10 LEADING:20 \
TRAILING:20 SLIDINGWINDOW:4:18 MINLEN:25
```

Trimmed reads were mapped using `bowtie/1.1.2` ([Langmead et al., 2009](<https://www.nature.com/articles/nmeth.1923>)) to the reference
genome `sacCer3/R64-1-1`. Three different mapping protocols were used.

# Lists  
## Bulleted List  

* Bullet 1  
* Bullet 2  
    + Sub-bullet 2.1  
    + Sub-bullet 2.2  
* Bullet 3 

# Images  
## Using Markdown  

![tiny-logo](./assets/logo_60.png)   

This image is 50% size.  
<img src="./assets/cover.jpg" style="width: 50%;" />

This image is 50% size, has shadow and corners rounded.  
<img src="./assets/cover.jpg" style="width: 50%;" class="fancyimage" />

This image is 25% size, has shadow and corners rounded.  
<img src="./assets/cover.jpg" style="width: 25%;" class="fancyimage" />

## Using R  

This image is restricted to a size of 50 pixels.  

```{r,echo=TRUE,fig.retina=NULL,out.width=50}
knitr::include_graphics("./assets/logo_60.png")
```

This image is restricted to a size of 300 pixels.  

```{r,echo=TRUE,fig.retina=NULL,out.width=300}
knitr::include_graphics("./assets/cover.jpg")
```

# R Output  

```{r,echo=TRUE}
data(iris)
str(iris)
```

# Slide with Math  

Some examples of rendering equations.

$e^{i\pi} + 1 = 0$
$$\frac{E \times X^2 \prod I}{2+7} = 432$$
$$\sum_{i=1}^n X_i$$
$$\int_0^{2\pi} \sin x~dx$$
$\left( \sum_{i=1}^{n}{i} \right)^2 = \left( \frac{n(n-1)}{2}\right)^2 = \frac{n^2(n-1)^2}{4}$
$\begin{eqnarray} X & \sim & \mathrm{N}(0,1)\\ Y & \sim & \chi^2_{n-p}\\ R & \equiv & X/Y \sim t_{n-p} \end{eqnarray}$
$\begin{eqnarray} P(|X-\mu| > k) & = & P(|X-\mu|^2 > k^2)\\ & \leq & \frac{\mathbb{E}\left[|X-\mu|^2\right]}{k^2}\\ & \leq & \frac{\mathrm{Var}[X]}{k^2} \end{eqnarray}$

# R tables  
## Paged  

View of the data using paged tables.

`r tc(name="table_paged",caption="_Table using paged tibble._")`  

```{r}
iris
```

## kable  

The most simple table using `kable` from R package `knitr`.

```{r}
knitr::kable(head(iris), 'html')
```

## kableExtra  

More advanced table using [`kableExtra`](https://cran.r-project.org/web/packages/kableExtra/vignettes/awesome_table_in_html.html) and [`formattable`](https://cran.r-project.org/web/packages/formattable/vignettes/formattable-data-frame.html).

`r tc(name="table_kableextra",caption="_Table using kableextra._")`  

```{r}
 iris[c(1:4,51:53,105:108),] %>%
  mutate(Sepal.Length=color_bar("lightsteelblue")(Sepal.Length)) %>%
  mutate(Sepal.Width=color_tile("white","orange")(Sepal.Width)) %>%
  mutate(Species=cell_spec(Species,"html",color="white",bold=T,
    background=c("#8dd3c7","#fb8072","#bebada")[factor(.$Species)])) %>%
  kable("html",escape=F) %>%
  kable_styling(bootstrap_options=c("striped","hover","responsive"),
                full_width=F,position="left") %>%
  column_spec(5,width="3cm")
```

## DT  

Interactive table using R package [`DT`](https://rstudio.github.io/DT/).

`r tc(name="table_datatable",caption="_Table using datatable._")`  

```{r}
datatable(iris,options=list(pageLength=7))
```

# Static Plots  
## Base Plot  

* Plots using base R are widely used and may be good enough for most situations. 
* But they lack a consistent coding framework.  

```{r,fig.height=6,fig.width=6}
{plot(x=iris$Sepal.Length,y=iris$Sepal.Width,
      col=c("coral","steelblue","forestgreen")[iris$Species],
      xlab="Sepal Length",ylab="Sepal Width",pch=19)
legend(x=7,y=4.47,legend=c("setosa","versicolor","virginica"),
       col=c("coral","steelblue","forestgreen"),pch=19)}
```

`r fc(name="static_base",caption="_Static plot using base plot._")`  

## ggplot2  

R package [`ggplot2`](http://ggplot2.org/) is the most versatile and complete plotting solution in any programming language ever.  

```{r,fig.height=6,fig.width=8}
iris %>%
  ggplot(aes(x=Sepal.Length,y=Sepal.Width,col=Species))+
  geom_point(size=2)+
  labs(x="Sepal Length",y="Sepal Width")+
  theme_report()
```

`r fc(name="static_ggplot",caption="_Static plot using ggplot2._")`  

## highcharter  

R package [`highcharter`](http://jkunst.com/highcharter/) is a wrapper around javascript library [`highcharts`](https://www.highcharts.com/).

```{r}
h <- iris %>%
  hchart(.,"scatter",hcaes(x="Sepal.Length",y="Sepal.Width",group="Species")) %>%
  hc_xAxis(title=list(text="Sepal Length"),crosshair=TRUE) %>%
  hc_yAxis(title=list(text="Sepal Width"),crosshair=TRUE) %>%
  hc_chart(zoomType="xy",inverted=FALSE) %>%
  hc_legend(verticalAlign="top",align="right") %>%
  hc_size(height=400)

htmltools::tagList(list(h))
```

`r fc(name="interactive_highcharter",caption="_Interactive scatterplot using highcharter._")`  

## plotly  

R package [`plotly`](https://plot.ly/r/) provides R binding around javascript plotting library [`plotly`](https://plot.ly).

```{r}
p <- iris %>% 
  plot_ly(x=~Sepal.Length,y=~Sepal.Width,color=~Species,width=500,height=400) %>% 
  add_markers()
p
```

`r fc(name="interactive_plotly",caption="_Interactive scatterplot using plotly._")`  

## ggplotly  

`plotly` also has a function called `ggplotly` which converts a static ggplot2 object into an interactive plot.

```{r,fig.height=5,fig.width=7}
p <- iris %>%
  ggplot(aes(x=Sepal.Length,y=Sepal.Width,col=Species))+
  geom_point()+
  labs(x="Sepal Length",y="Sepal Width")+
  theme_bw(base_size=12)

ggplotly(p,width=500,height=400)
```

`r fc(name="interactive_ggplotly",caption="_Interactive scatterplot using ggplotly._")`  

## ggiraph  

`ggiraph` is also an R package that can be used to convert a static ggplot2 object into an interactive plot.

```{r,fig.height=5,fig.width=7}
p <- ggplot(iris,aes(x=Sepal.Length,y=Petal.Length,colour=Species))+
      geom_point_interactive(aes(tooltip=paste0("<b>Petal Length:</b> ",Petal.Length,"\n<b>Sepal Length: </b>",Sepal.Length,"\n<b>Species: </b>",Species)),size=2)+
  theme_bw()

tooltip_css <- "background-color:#e7eef3;font-family:Roboto;padding:10px;border-style:solid;border-width:2px;border-color:#125687;border-radius:5px;"

ggiraph(code=print(p),hover_css="cursor:pointer;stroke:black;fill-opacity:0.3",zoom_max=5,tooltip_extra_css=tooltip_css,tooltip_opacity=0.9)
```

`r fc(name="interactive_ggiraph",caption="_Interactive scatterplot using ggiraph._")`  

## dygraphs  

R package [`dygraphs`](http://rstudio.github.io/dygraphs/) provides R bindings for javascript library [dygraphs](http://dygraphs.com/) for time series data.  

```{r,fig.height=5,fig.width=9}
lungDeaths <- cbind(ldeaths, mdeaths, fdeaths)
dygraph(lungDeaths,main="Deaths from Lung Disease (UK)") %>%
  dyOptions(colors=c("#66C2A5","#FC8D62","#8DA0CB"))
```

`r fc(name="interactive_dygraph",caption="_Interactive time series plot using dygraph._")`  

## Network Graph  

R package `networkD3` allows the use of interactive network graphs from the [D3.js](https://d3js.org/) javascript library.

```{r}
data(MisLinks,MisNodes)
forceNetwork(Links=MisLinks,Nodes=MisNodes,Source="source",
             Target="target",Value="value",NodeID="name",
             Group="group",opacity=0.4)
```

`r fc(name="interactive_network",caption="_Interactive network plot._")`  

## leaflet  

R package [`leaflet`](https://rstudio.github.io/leaflet/) provides R bindings for javascript mapping library; [leafletjs](http://leafletjs.com/).  

```{r}
leaflet(height=500,width=700) %>% 
  addTiles(urlTemplate='http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png') %>%
  #addProviderTiles(providers$Esri.NatGeoWorldMap) %>%
  addMarkers(lat=57.639327,lng=18.288534,popup="RaukR") %>%
  setView(lat=57.639327,lng=18.288534,zoom=15)
```

`r fc(name="interactive_map",caption="_Interactive map using leaflet._")`

## crosstalk  

R package [`crosstalk`](https://rstudio.github.io/crosstalk/index.html) allows `crosstalk` enabled plotting libraries to be linked. Through the shared 'key' variable, data points can be manipulated simultaneously on two independent plots.

```{r,fig.height=8,message=FALSE,warning=FALSE}
shared_quakes <- SharedData$new(quakes[sample(nrow(quakes), 100),])
lf <- leaflet(shared_quakes,height=300) %>%
        addTiles(urlTemplate='http://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png') %>%
        addMarkers()
py <- plot_ly(shared_quakes,x=~depth,y=~mag,size=~stations,height=300) %>% 
        add_markers()

htmltools::div(lf,py)
```

`r fc(name="interactive_crosstalk",caption="_Linking independent plots using crosstalk._")` 

# Directory  

```sh
.
+-- fq_raw/
+-- fq_trimmed/
+-- kallisto/
+-- mapping_star_genome/
+   +-- alignments/
+   +-- alignments_qc/
+   +-- featurecounts/
+   +-- qorts/
+   +-- rsem/
+-- qc/
+   +-- featurecounts_summary/
+   +-- fq_raw_qc
+   +-- fq_trimmed_logs
+   +-- fq_trimmed_qc
+-- salmon/
```

Total size is 70 GB.

# Session Info

* For more interactive graphics in R, go to <https://www.htmlwidgets.org/>
* For R Markdown, see <http://rmarkdown.rstudio.com> and <https://rmarkdown.rstudio.com/authoring_pandoc_markdown.html>

This document has been created in RStudio using R Markdown and related packages. For details about the OS, packages and versions, see detailed information below: 

```{r}
sessionInfo()
```

---

<div style="padding-bottom: 1.5em">
<span style="float:left; vertical-align:middle">
<b>`r format(Sys.Date(),format="%Y")`</b> | [SciLifeLab](https://www.scilifelab.se/) > [NBIS](https://nbis.se/) > [RaukR](https://nbisweden.github.io/workshop-RaukR-1806/)
</span>
<span style="float:right; vertical-align:middle">
<span class="footericon" style="padding-right:4px; padding-left:4px">
<a href="https://nbisweden.github.io/workshop-RaukR-1806/"><img src="./assets/icons8-globe-26.png" alt="website" border="0" style="height:15px"></a>
</span>
<span class="footericon" style="padding-right:4px; padding-left:4px">
<a href="https://twitter.com/hashtag/RaukR?src=hash"><img src="./assets/icons8-twitter-26.png" alt="twitter" border="0" style="height:15px"></a>
</span>
</span>
</div>
